Add 'crm sbd' sub-level (jsc#PED-8256)

liangxin1300 opened the pull request:
## Motivation
The main configurations for sbd use cases are scattered among sysconfig,
on-disk meta data, CIB, and even could be related to other OS components
eg. coredump, SCSI, multipath. 

It's desirable to reduce the management complexity among them and to
streamline the workflow for the main use case scenarios. 

## Changed include
#### Disk-based SBD scenarios

1. Show usage when syntax error
2. Completion
3. Display SBD related configuration (UC4 in PED-8256)
4. Change the on-disk meta data of the existing sbd disks (UC2.1 in PED-8256)
5. Add a sbd disk with the existing sbd configuration (UC2.2 in PED-8256)
6. Remove a sbd disk (UC2.3 in PED-8256)
7. Remove sbd from cluster
8. Replace the storage for a sbd disk (UC2.4 in PED-8256)]
9. display status (focusing on the runtime information only) (UC5 in PED-8256)

#### Disk-less SBD scenarios

1. Show usage when syntax error (diskless)
2. completion (diskless)
3. Display SBD related configuration (UC4 in PED-8256, diskless)
4. Manipulate the basic diskless sbd configuration (UC3.1 in PED-8256)

liangxin1300 commented:
<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->

- [Disk-based SBD scenarios](#disk-based-sbd-scenarios)
      + [1. Show usage when syntax error](#1-show-usage-when-syntax-error)
      + [2. Completion](#2-completion)
      + [4. Change the on-disk meta data of the existing sbd disks (UC2.1 in PED-8256)](#4-change-the-on-disk-meta-data-of-the-existing-sbd-disks-uc21-in-ped-8256)
      + [5. Add a sbd disk with the existing sbd configuration (UC2.2 in PED-8256)](#5-add-a-sbd-disk-with-the-existing-sbd-configuration-uc22-in-ped-8256)
      + [6. Remove a sbd disk (UC2.3 in PED-8256)](#6-remove-a-sbd-disk-uc23-in-ped-8256)
      + [7. Remove sbd from cluster](#7-remove-sbd-from-cluster)
      + [8. Replace the storage for a sbd disk (UC2.4 in PED-8256)](#8-replace-the-storage-for-a-sbd-disk-uc24-in-ped-8256)
- [Disk-less SBD scenarios](#disk-less-sbd-scenarios)

<!-- TOC end -->

<!-- TOC --><a name="disk-based-sbd-scenarios"></a>
## Disk-based SBD scenarios
- [1. Show usage when syntax error](#1-show-usage-when-syntax-error)
- [2. Completion](#2-completion)
- [3. Display SBD related configuration (UC4 in PED-8256)](#3-display-sbd-related-configuration-uc4-in-ped-8256)

<!-- TOC --><a name="1-show-usage-when-syntax-error"></a>
#### 1. Show usage when syntax error
```
# crm sbd configure xx
ERROR: Invalid argument: xx
Usage:
crm sbd configure show [disk_metadata|sysconfig|timeout]
crm sbd configure [device=<dev>]... [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
```
More syntax errror cases
See https://github.com/ClusterLabs/crmsh/pull/1491/commits/e91cc6ec92a12c7e4c9fdfe4c43925c5d72e324d
<!-- TOC --><a name="2-completion"></a>
#### 2. Completion
```
# crm sbd configure 
allocate-timeout=  device=            loop-timeout=      msgwait-timeout=   show               watchdog-timeout=  
     
# crm sbd configure device=/dev/sda5 
allocate-timeout=  device=            loop-timeout=      msgwait-timeout=   watchdog-timeout=

# crm sbd configure device=/dev/sda5 watchdog-timeout=5 
allocate-timeout=  device=            loop-timeout=      msgwait-timeout=

# crm sbd configure show 
disk_metadata   sysconfig       timeout
```
3. Display SBD related configuration (UC4 in PED-8256)
```
# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 7f04f327-3073-4121-a0bf-ce3d48723151
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure show sysconfig 
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=71
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=15
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=
SBD_DEVICE=/dev/sda5
```
<!-- TOC --><a name="4-change-the-on-disk-meta-data-of-the-existing-sbd-disks-uc21-in-ped-8256"></a>
#### 4. Change the on-disk meta data of the existing sbd disks (UC2.1 in PED-8256)
```
# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 20b85d71-005b-4198-b204-0767d301d164
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure watchdog-timeout=30
INFO: Initialize sbd device /dev/sda5
INFO: Sync /etc/sysconfig/sbd to all nodes
INFO: Requires to restart cluster service to take effect

# crm sbd configure show disk_metadata |grep watchdog
Timeout (watchdog) : 30
```
<!-- TOC --><a name="5-add-a-sbd-disk-with-the-existing-sbd-configuration-uc22-in-ped-8256"></a>
#### 5. Add a sbd disk with the existing sbd configuration (UC2.2 in PED-8256)
```
# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 14dce235-3386-4213-b531-b829ab7abab1
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE=/dev/sda5

# crm sbd configure device=/dev/sda7
INFO: Initialize sbd device /dev/sda7
INFO: Sync /etc/sysconfig/sbd to all nodes
INFO: Requires to restart cluster service to take effect

# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda7
Header version     : 2.1
UUID               : 12c511cd-f345-4729-8861-c82238b50508
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 5
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 10
==Header on disk /dev/sda7 is dumped
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 14dce235-3386-4213-b531-b829ab7abab1
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE="/dev/sda7;/dev/sda5"
```
<!-- TOC --><a name="6-remove-a-sbd-disk-uc23-in-ped-8256"></a>
#### 6. Remove a sbd disk (UC2.3 in PED-8256)
```
# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE="/dev/sda5;/dev/sda6"

# crm sbd remove /dev/sda6
INFO: Remove '/dev/sda6' from /etc/sysconfig/sbd
INFO: Sync /etc/sysconfig/sbd to all nodes
INFO: Requires to restart cluster service to take effect

# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE=/dev/sda5
```
<!-- TOC --><a name="7-remove-sbd-from-cluster"></a>
#### 7. Remove sbd from cluster
```
# crm sbd remove
INFO: Stop sbd resource stonith-sbd
INFO: Remove sbd resource 'stonith-sbd'
INFO: Disable sbd.service on node alp-1
WARNING: "stonith-enabled" in crm_config is set to false, it was true
INFO: Requires to restart cluster service to take effect
```
<!-- TOC --><a name="8-replace-the-storage-for-a-sbd-disk-uc24-in-ped-8256"></a>
#### 8. Replace the storage for a sbd disk (UC2.4 in PED-8256)
<!-- TOC --><a name="disk-less-sbd-scenarios"></a>
## Disk-less SBD scenarios

liangxin1300 commented:
<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->

- [Disk-based SBD scenarios](#disk-based-sbd-scenarios)
      + [1. Show usage when syntax error](#1-show-usage-when-syntax-error)
      + [2. Completion](#2-completion)
      + [4. Change the on-disk meta data of the existing sbd disks (UC2.1 in PED-8256)](#4-change-the-on-disk-meta-data-of-the-existing-sbd-disks-uc21-in-ped-8256)
      + [5. Add a sbd disk with the existing sbd configuration (UC2.2 in PED-8256)](#5-add-a-sbd-disk-with-the-existing-sbd-configuration-uc22-in-ped-8256)
      + [6. Remove a sbd disk (UC2.3 in PED-8256)](#6-remove-a-sbd-disk-uc23-in-ped-8256)
      + [7. Remove sbd from cluster](#7-remove-sbd-from-cluster)
      + [8. Replace the storage for a sbd disk (UC2.4 in PED-8256)](#8-replace-the-storage-for-a-sbd-disk-uc24-in-ped-8256)
- [Disk-less SBD scenarios](#disk-less-sbd-scenarios)

<!-- TOC end -->

<!-- TOC --><a name="disk-based-sbd-scenarios"></a>
## Disk-based SBD scenarios
- [1. Show usage when syntax error](#1-show-usage-when-syntax-error)
- [2. Completion](#2-completion)
- [3. Display SBD related configuration (UC4 in PED-8256)](#3-display-sbd-related-configuration-uc4-in-ped-8256)

<!-- TOC --><a name="1-show-usage-when-syntax-error"></a>
#### 1. Show usage when syntax error
```
# crm sbd configure xx
ERROR: Invalid argument: xx
Usage:
crm sbd configure show [disk_metadata|sysconfig|timeout]
crm sbd configure [device=<dev>]... [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
```
More syntax errror cases
See https://github.com/ClusterLabs/crmsh/pull/1491/commits/e91cc6ec92a12c7e4c9fdfe4c43925c5d72e324d
<!-- TOC --><a name="2-completion"></a>
#### 2. Completion
```
# crm sbd configure 
allocate-timeout=  device=            loop-timeout=      msgwait-timeout=   show               watchdog-timeout=  
     
# crm sbd configure device=/dev/sda5 
allocate-timeout=  device=            loop-timeout=      msgwait-timeout=   watchdog-timeout=

# crm sbd configure device=/dev/sda5 watchdog-timeout=5 
allocate-timeout=  device=            loop-timeout=      msgwait-timeout=

# crm sbd configure show 
disk_metadata   sysconfig       timeout
```
3. Display SBD related configuration (UC4 in PED-8256)
```
# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 7f04f327-3073-4121-a0bf-ce3d48723151
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure show sysconfig 
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=71
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=15
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=
SBD_DEVICE=/dev/sda5
```
<!-- TOC --><a name="4-change-the-on-disk-meta-data-of-the-existing-sbd-disks-uc21-in-ped-8256"></a>
#### 4. Change the on-disk meta data of the existing sbd disks (UC2.1 in PED-8256)
```
# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 20b85d71-005b-4198-b204-0767d301d164
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure watchdog-timeout=30
INFO: Initialize sbd device /dev/sda5
INFO: Sync /etc/sysconfig/sbd to all nodes
INFO: Requires to restart cluster service to take effect

# crm sbd configure show disk_metadata |grep watchdog
Timeout (watchdog) : 30
```
<!-- TOC --><a name="5-add-a-sbd-disk-with-the-existing-sbd-configuration-uc22-in-ped-8256"></a>
#### 5. Add a sbd disk with the existing sbd configuration (UC2.2 in PED-8256)
```
# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 14dce235-3386-4213-b531-b829ab7abab1
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE=/dev/sda5

# crm sbd configure device=/dev/sda7
INFO: Initialize sbd device /dev/sda7
INFO: Sync /etc/sysconfig/sbd to all nodes
INFO: Requires to restart cluster service to take effect

# crm sbd configure show disk_metadata 
==Dumping header on disk /dev/sda7
Header version     : 2.1
UUID               : 12c511cd-f345-4729-8861-c82238b50508
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 5
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 10
==Header on disk /dev/sda7 is dumped
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 14dce235-3386-4213-b531-b829ab7abab1
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped

# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE="/dev/sda7;/dev/sda5"
```
<!-- TOC --><a name="6-remove-a-sbd-disk-uc23-in-ped-8256"></a>
#### 6. Remove a sbd disk (UC2.3 in PED-8256)
```
# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE="/dev/sda5;/dev/sda6"

# crm sbd remove /dev/sda6
INFO: Remove '/dev/sda6' from /etc/sysconfig/sbd
INFO: Sync /etc/sysconfig/sbd to all nodes
INFO: Requires to restart cluster service to take effect

# crm sbd configure show sysconfig |grep DEVICE
SBD_DEVICE=/dev/sda5
```
<!-- TOC --><a name="7-remove-sbd-from-cluster"></a>
#### 7. Remove sbd from cluster
```
# crm sbd remove
INFO: Stop sbd resource stonith-sbd
INFO: Remove sbd resource 'stonith-sbd'
INFO: Disable sbd.service on node alp-1
WARNING: "stonith-enabled" in crm_config is set to false, it was true
INFO: Requires to restart cluster service to take effect
```
<!-- TOC --><a name="8-replace-the-storage-for-a-sbd-disk-uc24-in-ped-8256"></a>
#### 8. Replace the storage for a sbd disk (UC2.4 in PED-8256)
<!-- TOC --><a name="disk-less-sbd-scenarios"></a>
## Disk-less SBD scenarios

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,71 @@ Example:
+Usage:
+...............
+remove
+remove [<device> ...]
```
```suggestion
remove [<device=> ...]
```

zzhou1 commented:
When converting the disk-based SBD to the diskless SBD, fence_sbd primitive should be removed too.

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,71 @@ Example:
+
+# For disk-less SBD
+crm sbd configure show [sysconfig|property]
+crm sbd configure device="" [watchdog-device=<dev>] [watchdog-timeout=<integer>]
```
When converting the disk-based SBD to the diskless SBD, fence_sbd primitive should be removed too. Currently not yet.

nicholasyang2022 commented:
I played disk-based sbd and found some bugs and UE problems:

## show

Better to add section headers `sysconfig` and `property`.

## add a new disk

`crm sbd configure device=/dev/sda7` looks like replacing an existing device instead of adding a new device.

## remove a disk

Not working:

```
suse@ha-3-1:~> sudo crm sbd remove /dev/disk/by-path/ip-192.168.122.33:3260-iscsi-iqn.2024-09.com.example:faeb6079b4df45716081-lun-0 
ERROR: Device /dev/disk is not in config
```

```
suse@ha-3-1:~> sudo crm sbd remove device=/dev/disk/by-path/ip-192.168.122.33\:3260-iscsi-iqn.2024-09.com.example\:faeb6079b4df45716081-lun-0
ERROR: Device /dev/disk/by is not in config
```

And completion not working:

```
suse@ha-3-1:~> sudo crm sbd remove device=
.ansible/      .bash_history  .bashrc        bin/           .cache/        .config/       .emacs         .inputrc       .local/        .profile       .ssh/ 
```

And it is bad to have 2 different commands to do the same thing.

nicholasyang2022 commented:
## crm sbd status

Readability is not so good. `YES` and `NO` are confusing at 1st sight. Better to use TAB to align the fields.

```
> sudo crm sbd status
sbd.service status: (active|enabled|since)
ha-3-1: YES |NO  |active since: Tue 2024-09-10 16:26:31 CST
ha-3-2: YES |NO  |active since: Tue 2024-09-10 16:26:31 CST

watchdog info: (device|driver|kernel timeout)
ha-3-1: /dev/watchdog|iTCO_wdt|10
ha-3-2: /dev/watchdog|iTCO_wdt|10
```

## modify properties

Fore diskless use case, why do I have to specify `device=` when trying to modify properties?

```
> crm sbd configure watchdog-timeout=30
ERROR: No device specified
```

liangxin1300 commented:
> Better to add section headers `sysconfig` and `property`.

```
# crm sbd configure show
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 9a27dda6-2ba5-42aa-8db7-67c650179261
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped


==Content of /etc/sysconfig/sbd:
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=71
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=15
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=
SBD_DEVICE=/dev/sda5

==SBD related properties:
pcmk_delay_max=30s
have-watchdog=true
stonith-enabled=true
stonith-timeout=83
priority-fencing-delay=60
TimeoutStartUSec=90
```

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,482 @@
+            content_list = [line.strip() for line in f.readlines()
+                            if not line.startswith("#")
+                            and line.strip()]
+        for line in content_list:
```
Add such section header like `==Content of /etc/sysconfig/sbd:`

zzhou1 replied:
TimeoutStartUSec=90

is a different type

liangxin1300 replied:
> INFO: systemctl cat sbd.service

This command should be `systemctl show -p TimeoutStartUSec sbd --value`

> TimeoutStartUSec=90

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,482 @@
+    def _show_disk_metadata(self) -> None:
+        '''
+        Show sbd disk metadata for each configured device
+        '''
```
See above comment

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,71 @@ Example:
+[[cmdhelp.sbd.status,show SBD status]]
+==== `status`
+
+Show the status of the SBD daemon.
```
Buggy when the cluster is in the maintenance mode,

```
tws-corosync3-2:~ # crm_mon -rR1

Full List of Resources:
  * stonith-sbd (stonith:fence_sbd):     Stopped (maintenance)

tws-corosync3-2:~ # crm sbd status
fence_sbd status: 
(Empty result)
```

liangxin1300 replied:
> Buggy when the cluster is in the maintenance mode,
> 
> ```
> tws-corosync3-2:~ # crm_mon -rR1
> 
> Full List of Resources:
>   * stonith-sbd (stonith:fence_sbd):     Stopped (maintenance)
> 
> tws-corosync3-2:~ # crm sbd status
> fence_sbd status: 
> (Empty result)
> ```

That's because crm_resource direct info into stderr when resource NOT running
```
# crm_resource --locate --resource stonith-sbd 1> /dev/null
resource stonith-sbd is NOT running
```
I will try to catch both stdout and stderr here

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,482 @@
+                return
+            cmd = f"CIB_file={cib_path} crm configure show"
+        out = self.cluster_shell.get_stdout_or_raise_error(cmd)
+
```
> Sounds to me exit with ERROR is a too strong condition check.
tws-corosync3-2:~ # crm sbd configure device="/dev/disk/by-partlabel/sbd-tws-c3;/dev/disk/by-partlabel/sbd-tws-c2"
ERROR: pacemaker.service is not active

This ERROR case does make sense, `sbd configure <dev>` does depend on a running cluster

zzhou1 replied:
I do see my point as the future improvement, and do not have to be fixed by now.
And I understand your point as well.

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,71 @@ Example:
+...............
+# For disk-based SBD
+crm sbd configure show [disk_metadata|sysconfig|property]
+crm sbd configure [device=<dev>]... [watchdog-device=<dev>] [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
```
Some description to set up the proper user expectation. Something like,

```suggestion
crm sbd configure [device=<dev>]... [watchdog-device=<dev>] [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
        NOTE: 
        1. for the new device, it copies the disk metadata from the first existing SBD disk device.
        2. for the existing "device=<dev>", it will skip disk metadata operation and if no change.                       
```

nicholasyang2022 replied:
Why bother to write such complex document and expect users to read them? This can be avoided by better UI design. For example, we can reject `configure device=...` when a disk exists and require using `configure device+=...` to add a new disk.

zzhou1 replied:
> Why bother to write such complex document and expect users to read them? This can be avoided by better UI design. For example, we can reject `configure device=...` when a disk exists and require using `configure device+=...` to add a new disk.

It all about setting up the proper user expectation. I would try to rewrite a bit. Or, we have the better place or the better UI to deliver such expecation.

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,71 @@ Example:
+...............
+# For disk-based SBD
+crm sbd configure show [disk_metadata|sysconfig|property]
+crm sbd configure [device=<dev>]... [watchdog-device=<dev>] [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
```
Could have improvement for the disk-based sbd scenarios, there is no point to
" Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd "

```
tws-corosync3-2:~ # crm sbd configure device=/dev/vdb28 watchdog-timeout=2
INFO: Initializing SBD device /dev/vdb28
INFO: Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd: 2
```

liangxin1300 replied:
Here I want to log any value changed in /etc/sysconfig/sbd
"Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd" was corresponding to the given args `watchdog-timeout=2`
Does it make no sense?

liangxin1300 replied:
> For the disk-based sbd:
> 
> 1. I do suggest to delete such information which is misleading a little.
>    > INFO: Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd: 2
> 2. and, I event suggest to comment it out in /etc/sysconfig/sbd, eg.
>    #SBD_WATCHDOG_TIMEOUT=5

That means SBD_WATCHDOG_TIMEOUT in sysconfig/sbd is useless?

nicholasyang2022 replied:
And I am confused by `crm sbd configure watchdog-timeout=...`. There are 3 similar items: `Timeout (watchdog) :`, `SBD_WATCHDOG_TIMEOUT=` and `stonith-watchdog-timeout=`. Which ones are expected be modified by this command?

If `Timeout (watchdog) :` and `SBD_WATCHDOG_TIMEOUT=` controls the same thing, and only one of them is effective, we should show only the effective one in `crm sbd configure show`, or indicate which one is effective in some way.

More generally, the information given by `crm sbd configure show` is too low-level. It just dumps a lot of raw configuration. It is not symmetrical with command `crm sbd configure <name>=<value>`. As a pair of _getter_ and _setter_ command, I expected `crm sdb configure show` to give me back what I specified in `crm configure <name>=<value>`.

nicholasyang2022 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,482 @@
+                return self._configure_diskless(parameter_dict)
+
+        except self.SyntaxError as e:
+            logger.error(str(e))
```
```suggestion
            logger.error('%s', e)
```

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,71 @@ Example:
+...............
+# For disk-based SBD
+crm sbd configure show [disk_metadata|sysconfig|property]
+crm sbd configure [device=<dev>]... [watchdog-device=<dev>] [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
```
BUG: need exit immediately when choose "n" to reject overwriting sbd disk, otherwise it can create the inconsistent metadata, which is not supported.

```
tws-corosync3-2:~ # crm sbd configure device=/dev/vdb29
/dev/vdb29 has already been initialized by SBD, do you want to overwrite it (y/n)? n
INFO: Update SBD_DEVICE in /etc/sysconfig/sbd: /dev/vdb28;/dev/vdb29

tws-corosync3-2:~ # crm sbd configure show disk_metadata
INFO: crm sbd configure show disk_metadata
==Dumping header on disk /dev/vdb28
Header version     : 2.1
UUID               : 192cc144-37f5-4441-9d59-d547a30fcfae
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 2
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 10
==Header on disk /dev/vdb28 is dumped

==Dumping header on disk /dev/vdb29
Header version     : 2.1
UUID               : a4f362d9-454f-4221-9053-b344f9a1dcde
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 20
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/vdb29 is dumped

```

liangxin1300 replied:
> BUG: need exit immediately when choose "n" to reject overwriting sbd disk, otherwise it can create the inconsistent metadata, which is not supported.

This is our current logic, see the process to configure sbd in bootstrap stage, interactive mode
```
15sp6-1:~ # crm cluster init sbd
INFO: Loading "default" profile from /etc/crm/profiles.yml
WARNING: chronyd.service is not configured to start at system boot.
Do you want to continue anyway (y/n)? y
/etc/pacemaker/authkey already exists - overwrite (y/n)? y
INFO: Configure SBD:
  If you have shared storage, for example a SAN or iSCSI target,
  you can use it avoid split-brain scenarios by configuring SBD.
  This requires a 1 MB partition, accessible to all nodes in the
  cluster.  The device path must be persistent and consistent
  across all nodes in the cluster, so /dev/disk/by-id/* devices
  are a good choice.  Note that all data on the partition you
  specify here will be destroyed.

Do you wish to use SBD (y/n)? y
Path to storage device (e.g. /dev/disk/by-id/...), or "none" for diskless sbd, use ";" as separator for multi path []/dev/sda5;/dev/sda6
SBD is already configured to use /dev/sda5 - overwrite (y/n)? n
SBD is already configured to use /dev/sda6 - overwrite (y/n)? n
INFO: Restarting cluster service
INFO: BEGIN Waiting for cluster
...........                                                                                                                                                                                        INFO: END Waiting for cluster
WARNING: "stonith-enabled" in crm_config is set to true, it was false
INFO: BEGIN Adjusting sbd related timeout values
WARNING: "stonith-timeout" in crm_config is set to 71, it was 43
INFO: END Adjusting sbd related timeout values
INFO: Done (log saved to /var/log/crmsh/crmsh.log)
15sp6-1:~ # sbd -d /dev/sda5 dump
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 5d38e880-e644-4648-93b1-788f744d5559
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped
15sp6-1:~ # sbd -d /dev/sda6 dump
==Dumping header on disk /dev/sda6
Header version     : 2.1
UUID               : 96029eac-a917-430a-8670-9661cdadc304
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 5
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 10
==Header on disk /dev/sda6 is dumped
15sp6-1:~ # cat /etc/sysconfig/sbd |grep DEVICE
# SBD_DEVICE specifies the devices to use for exchanging sbd messages
#SBD_DEVICE=""
SBD_DEVICE="/dev/sda5;/dev/sda6"
```

Also need to change here?

zzhou1 replied:
In this process, sbd healthcheck logic should be useful and important. It must return the critical error in this situation in my opinion, and tell the user how to fix it. What would be the idea then?

crm cluster health sbd --fix
crm sbd device copy dev1 dev2

any other idea?

liangxin1300 replied:
> In this process, sbd healthcheck logic should be useful and important. It must return the critical error in this situation in my opinion, and tell the user how to fix it. What would be the idea then?
> 
> crm cluster health sbd --fix crm sbd device copy dev1 dev2
> 
Are we going to add `crm sbd device` interface this time?

> any other idea?

zzhou1 replied:
FYI, I don't have other elegant approach to copy dev1 to dev2, except "dd if=dev1 of=dev2 bs=XXX count=1". XXX usually is 512, but can be different, for example, in s390 as said in the man page.

nicholasyang2022 replied:
I think if users reject to overwrite the metadata, initialization process should abort immediately. If `crm cluster init sbd` does not do that in the past, it should be changed.

Think about in which case a user would reject to overwrite the metadata. Maybe a wrong device is specified. It does not make sense to continue setting up.

nicholasyang2022 replied:
> FYI, I don't have other elegant approach to copy dev1 to dev2, except "dd if=dev1 of=dev2 bs=XXX count=1". XXX usually is 512, but can be different, for example, in s390 as said in the man page.

Maybe we can read it from `/sys/class/block/<device>/queue/logical_block_size`?

zzhou1 replied:
> Think about in which case a user would reject to overwrite the metadata. Maybe a wrong device is specified.
I remember I heard the corner case that a user want to keep the disk metadata unchanged and reuse it.

zzhou1 replied:
>> FYI, I don't have other elegant approach to copy dev1 to dev2, except "dd if=dev1 of=dev2 bs=XXX count=1". XXX usually is 512, but can be different, for example, in s390 as said in the man page.
> Maybe we can read it from /sys/class/block/<device>/queue/logical_block_size?

The better interface is `strace blockdev --getss /dev/vdb28 2>&1 | grep ioctl`, for example. It calls ` ioctl(3, BLKSSZGET, [512])` and matches to the SBD code [1]


[1]: https://github.com/ClusterLabs/sbd/blob/cf5c2208bad2db2dff9b09624b89b05415c3bc11/src/sbd-md.c#L135

liangxin1300 replied:
> > Think about in which case a user would reject to overwrite the metadata. Maybe a wrong device is specified.
> 
> I remember I heard the corner case that a user want to keep the disk metadata unchanged and reuse it.

https://bugzilla.suse.com/show_bug.cgi?id=1201428

liangxin1300 commented:
> > Better to add section headers `sysconfig` and `property`.
> 
> ```
> # crm sbd configure show
> ==Dumping header on disk /dev/sda5
> Header version     : 2.1
> UUID               : 9a27dda6-2ba5-42aa-8db7-67c650179261
> Number of slots    : 255
> Sector size        : 512
> Timeout (watchdog) : 15
> Timeout (allocate) : 2
> Timeout (loop)     : 1
> Timeout (msgwait)  : 30
> ==Header on disk /dev/sda5 is dumped
> 
> 
> ==Content of /etc/sysconfig/sbd:
> SBD_PACEMAKER=yes
> SBD_STARTMODE=always
> SBD_DELAY_START=71
> SBD_WATCHDOG_DEV=/dev/watchdog0
> SBD_WATCHDOG_TIMEOUT=15
> SBD_TIMEOUT_ACTION=flush,reboot
> SBD_MOVE_TO_ROOT_CGROUP=auto
> SBD_SYNC_RESOURCE_STARTUP=yes
> SBD_OPTS=
> SBD_DEVICE=/dev/sda5
> 
> ==SBD related properties:
> pcmk_delay_max=30s
> have-watchdog=true
> stonith-enabled=true
> stonith-timeout=83
> priority-fencing-delay=60
> TimeoutStartUSec=90
> ```

Update:
```
# crm sbd configure show
INFO: crm sbd configure show disk_metadata
==Dumping header on disk /dev/sda6
Header version     : 2.1
UUID               : df9ae774-4f5b-4901-b690-b80099fd3bc8
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda6 is dumped


INFO: crm sbd configure show sysconfig
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=71
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=15
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=
SBD_DEVICE=/dev/sda6

INFO: crm sbd configure show property
pcmk_delay_max=30s
have-watchdog=true
stonith-enabled=true
stonith-timeout=83
priority-fencing-delay=60
stonith-watchdog-timeout=-1

INFO: systemctl show -p TimeoutStartUSec sbd --value
TimeoutStartUSec=90
```

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,482 @@
+            update_dict["SBD_WATCHDOG_DEV"] = watchdog_device
+        timeout_dict = {k: v for k, v in parameter_dict.items() if k in self.TIMEOUT_TYPES}
+
+        all_device_list = list(
+            dict.fromkeys(self.device_list_from_config + device_list)
+        )
```
To remove possible duplicated items while keeping device order unchanged

nicholasyang2022 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,488 @@
+    dev_char_class = r'[\w/\d;\-:.]+'
+    PARSE_RE = re.compile(
+		# Match "device" key with any value, including empty
+        fr'(device)=("[^"]*"|{dev_char_class})'
```
Diskless sbd does not works:

```
> sudo crm sbd configure device=""
ERROR: Invalid argument: device=
```

```
> sudo crm sbd configure 'device=""'
ERROR: Invalid argument: device=
```

liangxin1300 commented:
> ## crm sbd status
> Readability is not so good. `YES` and `NO` are confusing at 1st sight. Better to use TAB to align the fields.
> 
> ```
> > sudo crm sbd status
> sbd.service status: (active|enabled|since)
> ha-3-1: YES |NO  |active since: Tue 2024-09-10 16:26:31 CST
> ha-3-2: YES |NO  |active since: Tue 2024-09-10 16:26:31 CST
> 
> watchdog info: (device|driver|kernel timeout)
> ha-3-1: /dev/watchdog|iTCO_wdt|10
> ha-3-2: /dev/watchdog|iTCO_wdt|10
> ```


Now the output:
```
# crm sbd status
# Status of sbd.service:
Node   |Active  |Enabled |Since
alp-1  |YES     |YES     |active since: Wed 2024-09-18 15:53:55 CST
alp-2  |YES     |YES     |active since: Wed 2024-09-18 15:53:55 CST

# Watchdog info:
Node   |Device          |Driver    |Kernel Timeout
alp-1  |/dev/watchdog0  |iTCO_wdt  |10
alp-2  |/dev/watchdog0  |iTCO_wdt  |10

# Status of fence_sbd:
resource stonith-sbd is running on: alp-1
```

liangxin1300 commented:
> 
> ## modify properties
> Fore diskless use case, why do I have to specify `device=` when trying to modify properties?
> 
> ```
> > crm sbd configure watchdog-timeout=30
> ERROR: No device specified
> ```

Changed, no need to specify `device=` when diskless sbd is running

nicholasyang2022 reviewed (commented):
1. When I run `crm cluster init sbd -s ...` in a already initialized cluster, it ask me for overwriting `/etc/pacemaker/authkey`. This should not happen.

    ```
    sudo crm cluster init sbd -s /dev/disk/by-path/ip-192.168.122.33\:3260-iscsi-iqn.2024-09.com.example\:7abe4c1d432868b9b061-lun-0
    INFO: Loading "default" profile from /etc/crm/profiles.yml
    INFO: Loading "knet-default" profile from /etc/crm/profiles.yml
    /etc/pacemaker/authkey already exists - overwrite (y/n)?
    ```

2. Inconsistence in _"require restart"_ messages:

    For `crm sbd device add`:

    ```
    WARNING: Resource is running, need to restart cluster service manually on each node
    ```

    For `crm sbd disable`:

    ```
    INFO: Requires to restart cluster service to take effect
    ```

3. In failure scenarios, when a node is unreachable, `crm sbd status` breaks:

    ```
    suse@ha-3-1:~> sudo crm sbd status
    # Type of SBD:
    Disk-based SBD configured

    # Status of sbd.service:
    Node    |Active  |Enabled |Since
    ha-3-1  |YES     |YES     |active since: Thu 2024-10-24 16:57:13 CST
    ha-3-2  |YES     |YES     |active since: Thu 2024-10-24 16:57:14 CST
    ERROR: sbd.status: Failed to run command systemctl is-active 'sbd.service' on suse@ha-3-3: ssh: connect to host ha-3-3 port 22: No route to host Please configure passwordless authentication with "crm cluster init ssh" and "crm cluster join ssh"
    ```

liangxin1300 commented:
> 1. When I run `crm cluster init sbd -s ...` in a already initialized cluster, it ask me for overwriting `/etc/pacemaker/authkey`. This should not happen.
>    ```
>    sudo crm cluster init sbd -s /dev/disk/by-path/ip-192.168.122.33\:3260-iscsi-iqn.2024-09.com.example\:7abe4c1d432868b9b061-lun-0
>    INFO: Loading "default" profile from /etc/crm/profiles.yml
>    INFO: Loading "knet-default" profile from /etc/crm/profiles.yml
>    /etc/pacemaker/authkey already exists - overwrite (y/n)?
>    ```

Changed in #1598 
> 2. Inconsistence in _"require restart"_ messages:
>    For `crm sbd device add`:
>    ```
>    WARNING: Resource is running, need to restart cluster service manually on each node
>    ```
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    For `crm sbd disable`:
>    ```
>    INFO: Requires to restart cluster service to take effect
>    ```
Changed
> 3. In failure scenarios, when a node is unreachable, `crm sbd status` breaks:
>    ```
>    suse@ha-3-1:~> sudo crm sbd status
>    # Type of SBD:
>    Disk-based SBD configured
>    
>    # Status of sbd.service:
>    Node    |Active  |Enabled |Since
>    ha-3-1  |YES     |YES     |active since: Thu 2024-10-24 16:57:13 CST
>    ha-3-2  |YES     |YES     |active since: Thu 2024-10-24 16:57:14 CST
>    ERROR: sbd.status: Failed to run command systemctl is-active 'sbd.service' on suse@ha-3-3: ssh: connect to host ha-3-3 port 22: No route to host Please configure passwordless authentication with "crm cluster init ssh" and "crm cluster join ssh"
>    ```
Changed

nicholasyang2022 commented on crmsh/sh.py:
```diff
@@ -385,6 +385,14 @@ def get_rc_stdout_stderr_without_input(self, host, cmd) -> typing.Tuple[int, str
         rc, stdout, stderr = self.get_rc_stdout_stderr_raw_without_input(host, cmd)
         return rc, Utils.decode_str(stdout).strip(), Utils.decode_str(stderr).strip()
 
+    def get_rc_output_without_input(self, host, cmd) -> typing.Tuple[int, str]:
```
Ah. We already have one above: `get_rc_and_error`

nicholasyang2022 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,524 @@
+            print("# Status of fence_sbd:")
+            sbd_id_list = self.crm_mon_xml_parser.get_resource_id_list_via_type(sbd.SBDManager.SBD_RA)
+            for sbd_id in sbd_id_list:
+                rc, output = self.cluster_shell.get_rc_and_error(None, None, f"crm resource status {sbd_id}", ignore_output=False)
```
You need the output here. Then, I think the previous version, using a separated interface, is better

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,528 @@
+            self._show_disk_metadata()
+            if self.device_list_from_config:
+                print()
+            SBD._show_sysconfig()
```
Here are the current behavior:
```
# crm sbd purge
INFO: Stop sbd resource 'stonith-sbd'(stonith:fence_sbd)
INFO: Remove sbd resource 'stonith-sbd'
INFO: Disable sbd.service on node alp-1
INFO: Disable sbd.service on node alp-2
INFO: Move /etc/sysconfig/sbd to /etc/sysconfig/sbd.bak on all nodes
INFO: Delete cluster property "stonith-timeout" in crm_config
INFO: Delete cluster property "priority-fencing-delay" in crm_config
WARNING: "stonith-enabled" in crm_config is set to false, it was true
INFO: Restarting cluster service
INFO: BEGIN Waiting for cluster
..........                                                                                                                                                                      INFO: END Waiting for cluster

# ll /etc/sysconfig/sbd
ls: cannot access '/etc/sysconfig/sbd': No such file or directory

# ll /etc/sysconfig/sbd.bak 
-rw-r--r-- 1 root root 4225 Nov 25 15:54 /etc/sysconfig/sbd.bak

# crm sbd configure show
ERROR: sbd.service is not active
```

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,528 @@
+            return False
+        if not sbd.SBDUtils.is_using_disk_based_sbd():
+            logger.error("Only works for disk-based SBD")
+            logger.info("Please use 'crm cluster init sbd' to configure disk-based SBD first")
```
Changed as:
```
INFO: Please use 'crm cluster init -s <dev1> [-s <dev2> [-s <dev3>]]' to configure disk-based SBD first
```
Since the given device path might be very long

zzhou1 commented on doc/crm.8.adoc:
```diff
@@ -2104,6 +2104,83 @@ Example:
+- Show configured disk metadata
+- Show contents of /etc/sysconfig/sbd
+- Show SBD related cluster properties
+- Update the existing parameters
```
``` suggestion
- Update the SBD related configuration parameters
```

zzhou1 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,526 @@
+
+        return timeout_dict
+
+    def _configure_diskbase(self, parameter_dict: dict):
```
Ideally, I expect to see msgwait, SBD_DELAY_START, and TimeoutStartSec get updated along with `crm sbd configure watchdog-timeout=50`, for example

liangxin1300 replied:
```
# crm sbd configure show
INFO: crm sbd configure show disk_metadata
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : bf9db259-b5b8-408b-b438-8b229f468107
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped


INFO: crm sbd configure show sysconfig
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=71
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=15
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=
SBD_DEVICE=/dev/sda5

INFO: crm sbd configure show property
pcmk_delay_max=30s
have-watchdog=true
stonith-enabled=true
stonith-timeout=83
priority-fencing-delay=60

INFO: systemctl show -p TimeoutStartUSec sbd.service --value
TimeoutStartUSec=90

# crm sbd configure watchdog-timeout=50
INFO: No msgwait timeout specified, use 2*watchdog timeout: 100
INFO: Configuring disk-based SBD
INFO: Initializing SBD device /dev/sda5
INFO: Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd: 50
INFO: Already synced /etc/sysconfig/sbd to all nodes
WARNING: Resource is running, need to restart cluster service manually on each node
INFO: Update SBD_DELAY_START in /etc/sysconfig/sbd: 141
INFO: Already synced /etc/sysconfig/sbd to all nodes
WARNING: "stonith-timeout" in crm_config is set to 167, it was 83

# crm sbd configure show
INFO: crm sbd configure show disk_metadata
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 6a296240-c820-4b13-b174-238b9e49e6d9
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 50
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 100
==Header on disk /dev/sda5 is dumped


INFO: crm sbd configure show sysconfig
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=141
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=50
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=
SBD_DEVICE=/dev/sda5

INFO: crm sbd configure show property
pcmk_delay_max=30s
have-watchdog=true
stonith-enabled=true
stonith-timeout=167
priority-fencing-delay=60

INFO: systemctl show -p TimeoutStartUSec sbd.service --value
TimeoutStartUSec=169
```

liangxin1300 commented:
Add output of sbd process in `sbd status`:
```
# crm sbd status
# Type of SBD:
Disk-based SBD configured

# Status of sbd.service:
Node   |Active  |Enabled |Since
alp-1  |YES     |YES     |active since: Fri 2024-11-29 10:13:38 CST
alp-2  |YES     |YES     |active since: Fri 2024-11-29 10:13:40 CST

# Status of sbd process:
├─9074: sbd: watcher: /dev/sda5 - slot: 0 - uuid: 50ef8942-52ca-4ad5-9f4e-19c53e72eaa8
├─9075: sbd: watcher: /dev/sda6 - slot: 0 - uuid: 50cbaee5-a8a2-4d31-ad39-5fde9b3c3f2e
├─9076: sbd: watcher: /dev/sda7 - slot: 0 - uuid: 87933c2f-4ec5-4d35-a2b9-63534d3db85b

# Watchdog info:
Node   |Device          |Driver    |Kernel Timeout
alp-1  |/dev/watchdog0  |iTCO_wdt  |10
alp-2  |/dev/watchdog0  |iTCO_wdt  |10

# Status of fence_sbd:
resource stonith-sbd is running on: alp-1
```

zzhou1 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,554 @@
+        # To keep the order of devices during removal
+        left_device_list = [dev for dev in self.device_list_from_config if dev not in devices_to_remove]
+        if len(left_device_list) == 0:
+            raise self.SyntaxError("Not allowed to remove all devices")
```
```suggestion
            raise self.SyntaxError("Not allowed to remove all devices. Run `crm cluster init sbd -S` to bootstrap the diskless-sbd")
```

Intentionally, not give "-F" directly here, to have user think this twice. We can debate this.

liangxin1300 replied:
This is just the interface to remove device, not the interface to configure diskless SBD

zzhou1 replied:
Or, how about,
"Not allowed to remove all devices. If you do want the diskless-sbd, please run `crm cluster init sbd -S`"

liangxin1300 replied:
"Not allowed to remove all devices" is good enough, we are not clear on what exactly user want, this is my AI complete me:
```
Not allowed to remove all devices. Please use 'crm sbd purge' to remove all devices
```
I think the above case might happen, user just want to purge sbd from cluster

liangxin1300 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,538 @@
+            print(f"{node:<{max_node_len}}|{is_active_str:<{status_len}}|{is_enabled_str:<{status_len}}|{since_str_prefix}: {since}")
+        print()
+
+    def _print_watchdog_info(self):
```
Changed as: (when cluster service is stopped at alp-1)
```
alp-1:~ # crm sbd status
# Status of sbd.service:
Node   |Active  |Enabled |Since
alp-1  |NO      |YES     |disactive since: Fri 2024-11-29 21:34:44 CST
alp-2  |YES     |YES     |active since: Fri 2024-11-29 21:32:01 CST

# Status of sbd process on alp-1:
No sbd process found

# Status of sbd process on alp-2:
├─8993 sbd: watcher: /dev/sda5 - slot: 1 - uuid: 47509131-0eaa-4c6f-80ed-721a6aaf028e

# Watchdog info:
Node   |Device          |Driver    |Kernel Timeout
alp-1  |N/A             |N/A       |N/A
alp-2  |/dev/watchdog0  |iTCO_wdt  |10

# Status of fence_sbd:
resource stonith-sbd is running on: alp-2
```

zzhou1 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,551 @@
+        for node in self.cluster_nodes:
+            out = self.cluster_shell.get_stdout_or_raise_error(scripts_in_shell, node)
+            if out:
+                print(f"# Status of sbd process on {node}:")
```
```suggestion
                print(f"# Status of the sbd disk watcher process on {node}:")
```
And, this information should no be printed out for the diskless-sbd.

zzhou1 commented on crmsh/ui_sbd.py:
```diff
@@ -0,0 +1,551 @@
+        )
+        sbd_manager.init_and_deploy_sbd()
+        
+    def _configure_diskless(self, parameter_dict: dict):
```
Same as the disk-based sbd, I expect to see TimeoutStartSec get updated along with `crm sbd configure watchdog-timeout=50`, for example.

Sounds like, the diskless bootstrap code doesn't do this too.

liangxin1300 replied:
Already changed, please have a try
```
# crm sbd  configure watchdog-timeout=100
INFO: Configuring diskless SBD
WARNING: Diskless SBD requires cluster with three or more nodes. If you want to use diskless SBD for 2-node cluster, should be combined with QDevice.
INFO: Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd: 100
INFO: Already synced /etc/sysconfig/sbd to all nodes
INFO: Restarting cluster service
INFO: BEGIN Waiting for cluster
...........                                                                                      INFO: END Waiting for cluster
INFO: Update SBD_DELAY_START in /etc/sysconfig/sbd: 211
INFO: Already synced /etc/sysconfig/sbd to all nodes
WARNING: "stonith-timeout" in crm_config is set to 251, it was 71

# crm sbd configure show
INFO: crm sbd configure show sysconfig
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=211
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=100
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=

INFO: crm sbd configure show property
have-watchdog=true
stonith-enabled=true
stonith-watchdog-timeout=-1
stonith-timeout=251

INFO: systemctl show -p TimeoutStartUSec sbd.service --value
TimeoutStartUSec=253
```

zzhou1 replied:
Still need the similar handling when the value is decreased. For example, 

```
# crm sbd  configure watchdog-timeout=100
# crm sbd  configure watchdog-timeout=10 # still, SBD_DELAY_START=211, stonith-timeout=251, TimeoutStartUSec=253
# crm sbd  configure watchdog-timeout=50 # still, SBD_DELAY_START=211, stonith-timeout=251, TimeoutStartUSec=253
```

Even, I expect to delete "/etc/systemd/system/sbd.service.d/sbd_delay_start.conf" when TimeoutStartUSec <= 90

liangxin1300 replied:
Take `SBD_DELAY_START` for an example, current our algo is, if expected_value > configured _value, update, see https://github.com/ClusterLabs/crmsh/blob/master/crmsh/sbd.py#L232-L235
Do you mean we changed as `if expected_value != configuted_value`, update?

zzhou1 replied:
Good argument! You help me to recall my vague memory, in the past we let the algorithm adjust the value only if it can go wrong badly, and leave the value by assuming they might be manually configured for reasons . 

I wonder if makes more sense to refactor and add a more complex flag to control the levels to adjust those related values. Em, yes, we need think twice. Keep debating ... no rush.

Apparently, some related improvements as below is reasonable to fix now. `crm sbd purge` should cleanup systemd drop-in files regardless if sbd is the last stonith device. Same for stonith-watchdog-timeout.
 
  /etc/systemd/system/sbd.service.d/sbd_delay_start.conf
  /run/systemd/system/sbd.service.d/sbd_delay_start_disabled.conf

liangxin1300 replied:
> Good argument! You help me to recall my vague memory, in the past we let the algorithm adjust the value only if it can go wrong badly, and leave the value by assuming they might be manually configured for reasons .
> 
> I wonder if makes more sense to refactor and add a more complex flag to control the levels to adjust those related values. Em, yes, we need think twice. Keep debating ... no rush.

Yeah, maybe raise an upstream issue or create a confluence/jira page?

liangxin1300 replied:
> Apparently, some related improvements as below is reasonable to fix now. `crm sbd purge` should cleanup systemd drop-in files regardless if sbd is the last stonith device. Same for stonith-watchdog-timeout.
> 
> /etc/systemd/system/sbd.service.d/sbd_delay_start.conf /run/systemd/system/sbd.service.d/sbd_delay_start_disabled.conf

Add commit: Dev: sbd: Remove sbd delay start related diretories when running sbd purge

zzhou1 reviewed (approved):
Great work!
